replicator/run.py

```diff
@@ -0,0 +1,161 @@
+            print('Provisioning catalog ...')
+            pdns_request('post', path='/zones', body={'name': catalog_domain, 'kind': 'SLAVE', 'masters': [master_ip]})
+            serials[catalog_domain] = 0  # default for a new slave zone
+
+        # Get catalog
+        # TODO Maybe this should only be done if catalog is unset or if serials[catalog_domain] == local_catalog_serial,
+        #  i.e. if the previous catalog was fully processed (no left-over todos). This would ensure that everybody
+        #  falls behind at the same rate if there is high load. Otherwise, the order is random, and certain domains
+        #  may end up having a disadvantage in AXFR scheduling. (Not sure if this is the case.)
+        #  On the other hand, if several changes are done quickly on the same domain (e.g. delete/recreate domain),
+        #  fetching the new catalog here allows skipping useless intermediate states. Also, in case processing somehow
+        #  is stuck because the catalog is invalid, always fetching a new catalog improves chances of recovery.
+        #  Also, if this is not always done, then long-running AXFRs may hold up new changes to the catalog, as we
+        #  would be waiting for large zone's serial to update until we retrieve a new catalog.
+        catalog = pdns_request('get', path=f'/zones/{catalog_domain_id}').json()
+        local_catalog_serial = catalog['serial']
+
+        # Update the catalog in the background if necessary
+        remote_catalog_serial = query_serial(catalog_domain)
+        if local_catalog_serial != remote_catalog_serial:
+            print(f'Queueing catalog AXFR from {local_catalog_serial} to {remote_catalog_serial} ...')
+            pdns_request('put', path=f'/zones/{catalog_domain_id}/axfr-retrieve')
+
+        # See if the last parsed catalog is the current one
+        if serials[catalog_domain] == local_catalog_serial:
+            print(f'Nothing to do (catalog {serials[catalog_domain]} with {len(serials)} zones up to date).')
+            sleep(1)
+            continue
+
+        # Parse current state of catalog zone and determine slave-side changes
+        catalog_map = {}
+        for rrset in catalog['rrsets']:
+            m = re.match(r'^(serial)?\.?([0-9a-f]{40}).zones.catalog.internal.$', rrset['name'])
+            if m is None:
+                continue
+
+            subname, m_unique = m.groups()
+
+            map = catalog_map.setdefault(m_unique, {})
+            if rrset['type'] == 'PTR' and subname is None:
+                map['name'] = rrset['records'][0]['content']
+            elif rrset['type'] == 'TXT' and subname == 'serial':
+                map['serial'] = int(rrset['records'][0]['content'][1:-1])
+            else:
+                raise Exception('Unexpected catalog rrset %s' % rrset)
+
+        if not all(len(map) == 2 for map in catalog_map.values()):
+            raise Exception('Catalog has a value of m_unique with inconsistent content')
```

Comment:
log warning instead of crashing -- ignore spurious maps

Reply:
done